plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataBuilder.java

```diff
@@ -0,0 +1,328 @@
+import java.util.Objects;
+import java.util.TreeSet;
+
+import javax.annotation.Nullable;
+
+import org.sonatype.nexus.repository.maven.MavenPath;
+import org.sonatype.nexus.repository.maven.MavenPath.Coordinates;
+import org.sonatype.nexus.repository.maven.internal.maven2.metadata.MavenMetadata.Plugin;
+import org.sonatype.nexus.repository.maven.internal.maven2.metadata.MavenMetadata.Snapshot;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Function;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.eclipse.aether.util.version.GenericVersionScheme;
+import org.eclipse.aether.version.InvalidVersionSpecificationException;
+import org.eclipse.aether.version.Version;
+import org.eclipse.aether.version.VersionScheme;
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Maven 2 repository metadata builder.
+ *
+ * @since 3.0
+ */
+public class MetadataBuilder
+    extends ComponentSupport
+{
+  private final VersionScheme versionScheme;
+
+  private String groupId;
+
+  private String artifactId;
+
+  private String baseVersion;
+
+  public MetadataBuilder() {
+    this.versionScheme = new GenericVersionScheme();
+    // G
+    this.plugins = Lists.newArrayList();
+    // A
+    this.baseVersions = Sets.newTreeSet();
+    // V
+    this.latestVersionCoordinatesMap = Maps.newHashMap();
+  }
+
+  // -----------------------------------
+  // context
+
+  public String getGroupId() {
+    return groupId;
+  }
+
+  private boolean setGroupId(final String groupId) {
+    checkNotNull(groupId, "groupId");
+    if (Objects.equals(groupId, this.getGroupId())) {
+      return false;
+    }
+    this.groupId = groupId;
+    this.artifactId = null;
+    this.baseVersion = null;
+    return true;
+  }
+
+  public String getArtifactId() {
+    return artifactId;
+  }
+
+  private boolean setArtifactId(final String artifactId) {
+    checkState(groupId != null, "groupId == null");
+    checkNotNull(artifactId, "artifactId");
+    if (Objects.equals(artifactId, this.getArtifactId())) {
+      return false;
+    }
+    this.artifactId = artifactId;
+    this.baseVersion = null;
+    return true;
+  }
+
+  public String getBaseVersion() {
+    return baseVersion;
+  }
+
+  private boolean setBaseVersion(final String baseVersion) {
+    checkState(groupId != null, "groupId == null");
+    checkState(artifactId != null, "artifactId == null");
+    checkNotNull(baseVersion, "baseVersion");
+    if (Objects.equals(baseVersion, this.getBaseVersion())) {
+      return false;
+    }
+    this.baseVersion = baseVersion;
+    return true;
+  }
+
+  // -----------------------------------
+  // groupId
+
+  private final List<Plugin> plugins;
```

Comment:
I don't know if we explicitly have this as part of our code guidelines or not, but I would definitely prefer to have all of our fields defined consistently at the top of the file rather than peppered throughout the code